File: app/src/main/java/closer/vlllage/com/closer/ui/DragTriggerView.java
package closer.vlllage.com.closer.ui;

import android.support.annotation.NonNull;
import android.view.MotionEvent;
import android.view.View;

public class DragTriggerView {

    public DragTriggerView(@NonNull final View view, @NonNull OnDragEventListener onDragEventListener) {

        view.setClickable(true);
        view.setOnTouchListener(new View.OnTouchListener() {

            private TimedValue<Float> trackedX = new TimedValue<>(75);
            private TimedValue<Float> trackedY = new TimedValue<>(75);

            @Override
            public boolean onTouch(View view, MotionEvent motionEvent) {
                switch (motionEvent.getAction()) {
                    case MotionEvent.ACTION_DOWN:
                        trackedX.report(motionEvent.getRawX());
                        trackedY.report(motionEvent.getRawY());
                        onDragEventListener.onDragStart(trackedX, trackedY);
                        break;
                    case MotionEvent.ACTION_MOVE:
                        trackedX.report(motionEvent.getRawX());
                        trackedY.report(motionEvent.getRawY());
                        onDragEventListener.onDragUpdate(trackedX, trackedY);
                        break;
                    case MotionEvent.ACTION_UP:
                        trackedX.report(motionEvent.getRawX());
                        trackedY.report(motionEvent.getRawY());
                        onDragEventListener.onDragRelease(trackedX, trackedY);
                        break;
                }

                return true;
            }
        });
    }

    public interface OnDragEventListener {
        void onDragStart(TimedValue<Float> x, TimedValue<Float> y);
        void onDragRelease(TimedValue<Float> x, TimedValue<Float> y);
        void onDragUpdate(TimedValue<Float> x, TimedValue<Float> y);
    }
}

File: app/src/main/java/closer/vlllage/com/closer/ui/DragTriggerView.kt
package closer.vlllage.com.closer.ui

import android.view.MotionEvent
import android.view.View

class DragTriggerView(view: View, onDragEventListener: OnDragEventListener) {

    init {

        view.isClickable = true
        view.setOnTouchListener(object : View.OnTouchListener {

            private val trackedX = TimedValue<Float>(75)
            private val trackedY = TimedValue<Float>(75)

            override fun onTouch(view: View, motionEvent: MotionEvent): Boolean {
                when (motionEvent.action) {
                    MotionEvent.ACTION_DOWN -> {
                        trackedX.report(motionEvent.rawX)
                        trackedY.report(motionEvent.rawY)
                        onDragEventListener.onDragStart(trackedX, trackedY)
                    }
                    MotionEvent.ACTION_MOVE -> {
                        trackedX.report(motionEvent.rawX)
                        trackedY.report(motionEvent.rawY)
                        onDragEventListener.onDragUpdate(trackedX, trackedY)
                    }
                    MotionEvent.ACTION_UP -> {
                        trackedX.report(motionEvent.rawX)
                        trackedY.report(motionEvent.rawY)
                        onDragEventListener.onDragRelease(trackedX, trackedY)
                    }
                }

                return true
            }
        })
    }

    interface OnDragEventListener {
        fun onDragStart(x: TimedValue<Float>, y: TimedValue<Float>)
        fun onDragRelease(x: TimedValue<Float>, y: TimedValue<Float>)
        fun onDragUpdate(x: TimedValue<Float>, y: TimedValue<Float>)
    }
}
